File: fem/C++/openmp-target/manu.def/src/mat_ass_bc.c
/**
 ** MAT_ASS_BC
 **/
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "pfem_util.h"

extern FILE *fp_log;
void MAT_ASS_BC()
{
  /*
  int icel;
  int in1,in2,in3,in4,in5,in6,in7,in8;
  int iq1,iq2,iq3,iq4,iq5,iq6,iq7,iq8;
  int iS,iE;
  KREAL STRESS,VAL;
  */
  int i,j,k,in,ib,ib0;
  KINT *IWKX;
  
  IWKX=(KINT*)malloc(sizeof(KINT)*NP*2);
  for(i=0;i<NP;i++) for(j=0;j<2;j++) IWKX[i*2+j]=0;
  
/**
   Z=Zmax
**/
#pragma omp parallel for private(in)
  for(in=0;in<NP;in++) IWKX[in*2+0]=0;
  
  ib0=-1;
  for( ib0=0;ib0<NODGRPtot;ib0++){
    if( strcmp(NODGRP_NAME[ib0].name,"Zmax") == 0 ) break;
  }
#pragma omp parallel for private(ib,in)    
  for( ib=NODGRP_INDEX[ib0];ib<NODGRP_INDEX[ib0+1];ib++){
    in=NODGRP_ITEM[ib];
    IWKX[(in-1)*2+0]=1;
  }

#pragma omp target data \
  map(to: IWKX[0:NP*2],indexLU[0:NP+1],itemLU[0:NPLU])
  {
  
#pragma omp target teams distribute parallel for	   \
  map(alloc: IWKX[0:NP*2],indexLU[0:NP+1])		   \
  map(alloc: AMAT[0:NPLU],B[0:NP],D[0:NP])		   \
  private(in,k)
  for(in=0;in<NP;in++){
    if( IWKX[in*2+0] == 1 ){
      B[in]= 0.e0;
      D[in]= 1.e0;
      for(k=indexLU[in];k<indexLU[in+1];k++){
	AMAT[k]= 0.e0;
      }
    }
  }

#pragma omp target teams distribute parallel for \
  map(alloc: IWKX[0:NP*2],indexLU[0:NP+1],itemLU[0:NPLU])	\
  map(alloc: AMAT[0:NPLU]) \
  private(in,k)
  for(in=0;in<NP;in++){
    for(k=indexLU[in];k<indexLU[in+1];k++){
      if (IWKX[itemLU[k]*2+0] == 1 ) {
	AMAT[k]= 0.e0;
      }
    }
  }

  } // pragma omp target end data

  free(IWKX); 
}

File: fem/F/openmp-target/manu.def/src/mat_ass_bc.f
!C
!C***
!C*** MAT_ASS_BC
!C***
!C
      subroutine MAT_ASS_BC
      use pfem_util
      implicit none
      integer :: k,in,ib,ib0
      integer(kind=kint), dimension(:,:),allocatable :: IWKX

      allocate (IWKX(NP,2))
      IWKX= 0

!C
!C== Z=Zmax

!$omp parallel do private(in)
      do in= 1, NP
        IWKX(in,1)= 0
      enddo

      ib0= -1
      do ib0= 1, NODGRPtot
        if (NODGRP_NAME(ib0).eq.'Zmax') exit
      enddo

!$omp parallel do private(ib,in)      
      do ib= NODGRP_INDEX(ib0-1)+1, NODGRP_INDEX(ib0)
        in= NODGRP_ITEM(ib)
        IWKX(in,1)= 1
      enddo

!$omp target data
!$omp& map(to: IWKX(1:NP,1:2),indexLU(0:NP),itemLU(1:NPLU))

!$omp target teams distribute parallel do
!$omp& map(alloc: IWKX(1:NP,1:2),indexLU(0:NP))
!$omp& map(alloc: AMAT(1:NPLU),B(1:NP),D(1:NP))
!$omp& private(in,k)
      do in= 1, NP
        if (IWKX(in,1).eq.1) then
          B(in)= 0.d0
          D(in)= 1.d0
          do k= indexLU(in-1)+1, indexLU(in)
            AMAT(k)= 0.d0
          enddo
        endif
      enddo

!$omp target teams distribute parallel do
!$omp& map(alloc: IWKX(1:NP,1:2),indexLU(0:NP),itemLU(1:NPLU))
!$omp& map(alloc: AMAT(1:NPLU))
!$omp& private(in,k)
      do in= 1, NP
        do k=indexLU(in-1)+1,indexLU(in)
          if (IWKX(itemLU(k),1).eq.1) then
            AMAT(k)= 0.d0
          endif
        enddo
      enddo
!C==

!$omp end target data
      deallocate(IWKX)
      return
      end
